cpp/src/arrow/io/io-readahead-test.cc

```diff
@@ -0,0 +1,235 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include <chrono>
+#include <memory>
+#include <string>
+#include <thread>
+#include <vector>
+
+#include <gtest/gtest.h>
+
+#include "arrow/buffer.h"
+#include "arrow/io/interfaces.h"
+#include "arrow/io/memory.h"
+#include "arrow/io/readahead.h"
+#include "arrow/memory_pool.h"
+#include "arrow/status.h"
+#include "arrow/test-util.h"
+#include "arrow/util/checked_cast.h"
+
+namespace arrow {
+namespace io {
+namespace internal {
+
+static void sleep_for(double seconds) {
+  std::this_thread::sleep_for(
+      std::chrono::nanoseconds(static_cast<int64_t>(seconds * 1e9)));
+}
+
+static void busy_wait(double seconds, std::function<bool()> predicate) {
+  const double period = 0.001;
+  for (int i = 0; !predicate() && i * period < seconds; ++i) {
+    sleep_for(period);
+  }
+}
+
+std::shared_ptr<BufferReader> DataReader(const std::string& data) {
+  std::shared_ptr<Buffer> buffer;
+  ABORT_NOT_OK(AllocateBuffer(data.length(), &buffer));
+  memcpy(buffer->mutable_data(), data.data(), data.length());
+  return std::make_shared<BufferReader>(std::move(buffer));
+}
+
+static int64_t WaitForPosition(std::shared_ptr<RandomAccessFile> file, int64_t expected,
+                               double seconds = 0.2) {
+  int64_t pos = -1;
+  busy_wait(seconds, [&]() -> bool {
+    ABORT_NOT_OK(file->Tell(&pos));
+    return pos >= expected;
+  });
+  return pos;
+}
+
+static void AssertEventualPosition(std::shared_ptr<RandomAccessFile> file,
```

Comment:
Better to get in habit of passing `RandomAccessFile*` unless you need the `shared_ptr<T>`

Reply:
Indeed, not a big deal here, but in general utility code if a function doesn't need to return ownership of `shared_ptr<T>` then passing `T*` makes it more versatile

Reply:
Ok, updated.